src/core/qgsstoredexpressionmanager.cpp

```diff
@@ -0,0 +1,147 @@
+/***************************************************************************
+                          qgsstoredexpressionmanager.cpp
+                             -------------------
+    begin                : August 2019
+    copyright            : (C) 2019 David Signer
+    email                : david at opengis dot ch
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsstoredexpressionmanager.h"
+#include "qgis.h"
+
+
+#include <QDomElement>
+
+QString QgsStoredExpressionManager::addStoredExpression( const QString &name, const QString &expression, const QgsStoredExpression::Category &tag )
+{
+  Q_UNUSED( tag );
+
+  QgsStoredExpression storedExpression( name, expression, tag );
+
+  mStoredExpressions.append( storedExpression );
+
+  return storedExpression.id;
+}
+
+void QgsStoredExpressionManager::removeStoredExpression( const QString &id )
+{
+  int i = 0;
+  for ( const QgsStoredExpression &storedExpression : qgis::as_const( mStoredExpressions ) )
+  {
+    if ( storedExpression.id == id )
+    {
+      mStoredExpressions.removeAt( i );
+      //leave the loop after editing source
+      break;
+    }
+    ++i;
+  }
+}
+
+void QgsStoredExpressionManager::updateStoredExpression( const QString &id, const QString &name, const QString &expression, const QgsStoredExpression::Category &tag )
+{
+  Q_UNUSED( tag );
+
+  int i = 0;
+  for ( const QgsStoredExpression &storedExpression : qgis::as_const( mStoredExpressions ) )
+  {
+    if ( storedExpression.id == id )
+    {
+      QgsStoredExpression newStoredExpression = mStoredExpressions.at( i );
+      newStoredExpression.name = name;
+      newStoredExpression.expression = expression;
+      newStoredExpression.tag = tag;
+      mStoredExpressions.replace( i, newStoredExpression );
+      //leave the loop after editing source
+      break;
+    }
+    ++i;
+  }
+}
+
+void QgsStoredExpressionManager::addStoredExpressions( QList< QgsStoredExpression > storedExpressions )
+{
+  mStoredExpressions.append( storedExpressions );
+}
+
+QList< QgsStoredExpression > QgsStoredExpressionManager::storedExpressions( const QgsStoredExpression::Category &tag )
+{
+  Q_UNUSED( tag );
```

Comment:
I think this should be filtering for matching tags here, right?

Reply:
Yes, you're right.

Reply:
Sorry, I just see, this tag-integration has not been a very proper commit. I will update my tests for these cases as well.